### ISSUE TITLE
running client on OS X fails with permission errors

### ISSUE THREAD

**Issue by jmhodges** [letsencrypt/letsencrypt#680]
It seems that there is some python code that tries to test if a port is open which does not work unless running under sudo. (This also means the integration tests always fail on OS X machines.)

I believe that is being done in order to see if the client can later open that port itself. A nice solution would be to just try to open the port and then handle the failures there, instead. This has the nice property of removing a possible race between the time of port-check and the time of port-allocation where another service could take the port.

**Comment by jmhodges** [comment]
Oh, I forgot the stacktrace.

```
Traceback (most recent call last):
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/venv/bin/letsencrypt", line 9, in <module>
    load_entry_point('letsencrypt==0.1', 'console_scripts', 'letsencrypt')()
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/letsencrypt/cli.py", line 778, in main
    return args.func(args, config, plugins)
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/letsencrypt/cli.py", line 225, in auth
    domains, authenticator, installer, plugins):
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/letsencrypt/client.py", line 240, in obtain_and_enroll_certificate
    certr, chain, key, _ = self.obtain_certificate(domains)
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/letsencrypt/client.py", line 216, in obtain_certificate
    return self._obtain_certificate(domains, csr) + (key, csr)
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/letsencrypt/client.py", line 174, in _obtain_certificate
    authzr = self.auth_handler.get_authorizations(domains)
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/letsencrypt/auth_handler.py", line 79, in get_authorizations
    cont_resp, dv_resp = self._solve_challenges()
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/letsencrypt/auth_handler.py", line 113, in _solve_challenges
    dv_resp = self.dv_auth.perform(self.dv_c)
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/letsencrypt/plugins/standalone/authenticator.py", line 386, in perform
    if self.already_listening(self.config.dvsni_port):
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/letsencrypt/plugins/standalone/authenticator.py", line 304, in already_listening
    listeners = [conn.pid for conn in psutil.net_connections()
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/venv/lib/python2.7/site-packages/psutil/__init__.py", line 1742, in net_connections
    return _psplatform.net_connections(kind)
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/venv/lib/python2.7/site-packages/psutil/_psosx.py", line 160, in net_connections
    cons = Process(pid).connections(kind)
  File "/Users/jmhodges/src/github.com/letsencrypt/boulder/letsencrypt/venv/lib/python2.7/site-packages/psutil/_psosx.py", line 210, in wrapper
    raise AccessDenied(self.pid, self._name)
AccessDenied: psutil.AccessDenied (pid=97671)
```

**Comment by bmw** [comment]
Thanks @jmhodges. I have a Macbook back in Michigan I can use to take a look at this. I'll try to submit a PR to fix it sometime next week.

**Comment by bmw** [comment]
After taking a look at the client on OS X, it appears it needs some work and I'm not sure that I'm the best person do it. I'm removing my assignment to this for now.

**Comment by kuba** [comment]
@BKreisel, could you have a look at this one? :)

**Comment by BKreisel** [comment]
@kuba I'm looking into it. Putting in a flag to skip the check on OS X seems to work at the moment. I'm checking through the other binding code to look for implications I might have missed. I'm debugging another error that occurs later in the boulder integration script, but it doesn't seem related to this isuse.

**Comment by jmhodges** [comment]
Why would skip the check instead of just trying to grab the port as the check if it's just going to fail later anyway? This would work across OSes and not require duplicating the "I can't get the port" error code that has to exist later, anyhow.

I imagine that would require some refactoring, of course!

**Comment by BKreisel** [comment]
I agree, that sounds like a much better solution in the long run. I can't think of a reason offhand why the application would need to know in advance. There may also be an added benefit of eliminating some kind of unforseen port-related race conditions

**Comment by jdkasten** [comment]
I think the original thought was to do it in the prepare function so that it could notify the user that it was blocked or unavailable.  Which would be displayed to the user on the authenticator menu. 

The user could either cancel or make a better informed decision.  I think @kuba was working on making standalone more robust... so I don't think that this should be a blocker.

**Comment by bmw** [comment]
Fixed in #886.
